src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs

```diff
@@ -128,6 +129,16 @@ public class SessionStateTempDataProvider : ITempDataProvider
                             throw new InvalidOperationException(message);
                         }
                     }
+                    else if (item.Value is long)
```

Comment:
Meh. Unclear why we'd guess the original type was `int` and not `short` et cetera.